add optional annotations artifact with SingleIn and ForScope

gabrielittner opened the pull request:
Closes #692

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/anvil?pullRequest=725) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/anvil?pullRequest=725) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/anvil?pullRequest=725) it.</sub>

kingsleyadio commented:
(opinion): I think optional already has a meaning in Dagger world, which might make this naming a little confusing.
I propose changing for example, to something like `annotations-extended`. Other than that, love the reasoning behind including these directly in Anvil

gabrielittner commented:
I mostly went with the suggestion from the issue but don't really have a strong opinion on the artifact name. Happy to update the pr with something else

vRallev commented:
It would be great to update the commit message with a better description. I'm okay with `optional` as name, but what I'd like to see is a DSL option that adds this dependency for you, similar to how the Gradle plugin adds the annotations artifact by default. This can be a separate PR, but this one shouldn't resolve #692.

ZacSweers commented:
I think optional is a bad name because dagger has specialized support for an "optional" semantic via `@BindsOptionalOf`. I think scope-annotations or something would be better.

Also - does this need to be configurable via the DSL since it technically doesn't really have anything to do with anvil? 🤔

vRallev commented:
If it doesn't have anything to do with Anvil, then it shouldn't be part of the repository altogether. By allowing it to be configured through the DSL we don't have to deal with Maven coordinates and it's more discoverable IMO.

JoelWilcox commented:
I agree that having DSL support for configuring/adding this dependency would be nice. @gabrielittner up to you if you want to take a stab at that, if not I'm happy to add it separately.

JoelWilcox commented:
> I think optional is a bad name because dagger has specialized support for an "optional" semantic via @BindsOptionalOf. I think scope-annotations or something would be better.

Hmm this is a fair point, although I feel there aren't a lot of other name options that still accurately capture it as being a set of optional annotations while still being simple. I think something like `scope-annotations` could be confusing in other ways since the main annotations inherently deal with scopes too, and it could be limiting in the future if there are other optional opinionated annotations we want to add (which maybe don't directly deal with scopes).

At the moment I'd lean more towards some potential initial confusion with Dagger's concept of `Optional` since it should quickly become obvious it's unrelated when looking at the documentation for the artifact or directly at the included annotations. Only other name that comes to mind so far as a middle-ground would be something like `annotations-extra`.

gabrielittner commented:
@vRallev I've rebased and updated the commit message.

The DSL now has `addOptionalAnnotations` to add the dependency. It's also on `VariantFilter` to match the other options but if you think that's too much we can remove it from there and just directly use the main property from the extension.

JoelWilcox commented on gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilExtension.kt:
```diff
@@ -50,6 +50,13 @@ public abstract class AnvilExtension @Inject constructor(objects: ObjectFactory)
     .convention(false)
 
+  /**
+   * Add the `annotations-optional` artifact as a dependency to make `SingleIn` and `ForScope` 
+   * available to use.
```
Minor re-wording suggestion:
```suggestion
   * Add the `annotations-optional` artifact as a dependency to make annotations like `@SingleIn` and `@ForScope` 
   * available to use. These are annotations that are not strictly required but which we've found to be helpful with managing larger dependency graphs.
```

JoelWilcox commented on gradle-plugin/src/main/java/com/squareup/anvil/plugin/VariantFilter.kt:
```diff
@@ -36,6 +36,12 @@ public interface VariantFilter : Named {
   public var syncGeneratedSources: Boolean
+
+  /**
+   * Add the `annotations-optional` artifact as a dependency to make `SingleIn` and `ForScope` 
+   * available to use.
```
```suggestion
   * Add the `annotations-optional` artifact as a dependency. The default
   * value comes from the [AnvilExtension]. See [AnvilExtension.addOptionalAnnotations] for more
   * details.
```

gabrielittner commented on build.gradle:
```diff
@@ -94,6 +94,7 @@ subprojects {
+          && project.name != "annotations-optional"
           && project.name != "scopes"
           && project.name != "dagger-factories-only"
       if (optInExperimental) {
```
Added it [6f175ec](https://github.com/square/anvil/pull/725/commits/6f175ec5a64e55dbceab6c8f8345f7202b35288d)

JoelWilcox reviewed (approved):
Lgtm! Thanks for adding this

matejdro commented:
Could a mention of this be added to the `README`? Right now this is not very discoverable.
